earth_enterprise/src/fusion/autoingest/sysman/InsetTilespaceIndex.h

```diff
@@ -0,0 +1,190 @@
+
+#include <string>
+#include <functional>
+#include "common/khTileAddr.h"
+#include "common/quadtreepath.h"
+#include "common/khExtents.h"
+#include "common/khInsetCoverage.h"
+#include "fusion/autoingest/sysman/InsetInfo.h"
+#include <iostream>
+//#include "fusion/autoingest/sysman/InsetTilespaceIndex.h"
+
+// // TODO - pack this struct to fit in a tidy 64 bit "size_t" for optimal performance
+// struct  TilespaceMBR { 
+//     byte projectionType,  // TODO - Flat, Mercator, etc.  Reuse Define  a definition other headers?
+//     byte maxDepth,
+//     uint32 quadTreeBits
+// };
+template <class T>
+class QuadKeyTreeNode {
+//     friend class QuadKeyTree<T>;
+// protected:
+public:
+    //QuadKeyTreeNode() = default {}
+    std::array<std::shared_ptr<QuadKeyTreeNode<T>>, 4> nodes {{nullptr, nullptr, nullptr, nullptr}};
+    std::vector<const T*> data;
+};
+
+template <class T>
+class QuadKeyTree {
+private:
+    //std::array<std::shared_ptr<QuadKeyTreeNode<T>>, 4> nodes {{nullptr, nullptr, nullptr, nullptr}};
+    std::shared_ptr<QuadKeyTreeNode<T>> levelZeroNode = std::make_shared<QuadKeyTreeNode<T>>();
+public:
+    void AddElementAtQuadTreePath(QuadtreePath path,  const T *element){
+        const uint pathLength = path.Level();
+        uint pathIndex = 0;
+        std::shared_ptr<QuadKeyTreeNode<T>> nodeptr = levelZeroNode;
+
+        while (pathIndex < pathLength){
+            uint qtIndexAtLevel = path[pathIndex];     // e.g., if path is "1320", tileAtLevel would be 3 for level=2
+            if (nodeptr->nodes[qtIndexAtLevel] == nullptr)
+                nodeptr->nodes[qtIndexAtLevel] = std::make_shared<QuadKeyTreeNode<T>>();
+            nodeptr = nodeptr->nodes[qtIndexAtLevel];
+            pathIndex++;
+        }
+
+        assert(nodeptr != nullptr);
+        nodeptr->data.push_back(element);
+
+        // if (path.Level() == 0)
+        //     levelZeroNode->data.push_back(element);
+        // else{
+        //     if (levelZeroNode->nodes[path[0]] == nullptr)
+        //         levelZeroNode->nodes[path[0]] = std::make_shared<QuadKeyTreeNode<T>>();
+        //     std::shared_ptr<QuadKeyTreeNode<T>> nodeptr = levelZeroNode->nodes[path[0]];
+        //     for (uint32 level = 1; level < path.Level(); level++) {
+        //         if (nodeptr->nodes[path[level]] == nullptr)
+        //             nodeptr->nodes[path[level]] = std::make_shared<QuadKeyTreeNode<T>>();
+        //         nodeptr = nodeptr->nodes[path[level]];
+        //     }
+
+        //     assert(nodeptr != nullptr);
+        //     nodeptr->data.push_back(element);
+        // }
+    }
+
+    std::vector<const T*> GetElementsAtQuadTreePath(QuadtreePath path, uint startLevel) 
+    {
+        std::vector<const T*> vec;
+
+        const uint pathLength = path.Level();
+        uint pathIndex = 0;
+        std::shared_ptr<QuadKeyTreeNode<T>> nodeptr = levelZeroNode;
+
+        while (pathIndex < pathLength && pathIndex < startLevel && nodeptr != nullptr){
```

Comment:
Minor: it occurred to me that we might be able to save some time by returning two lists - a list of items "above" this one in the tree and a list of items "below" this one in the tree. I think the ones below are guaranteed to overlap while the ones above may or may not overlap. We could use that to reduce the number of comparisons even further.

I haven't thought this through all the way, so it might not work. In any case, it's something to worry about later.